### ISSUE TITLE
Add support for Map like scope objects

### ISSUE THREAD

**Issue by jhugman** [josdejong/mathjs#2143]
I'd like to be able to use `Map` as a scope object.

```javascript
const scope = new Map()
scope.set('a', 3)
scope.set('b', 5)

math.evaluate('a * b', scope)
```

Ideally, I'd like to be able to use a custom scope object:

```javascript
class FancyScopeObject {
   set(key, value) { … }
   get(key) { … }
}

const scope = new FancyScopeObject()
scope.set('a', 3)
scope.set('b', 5)

math.evaluate('a * b', scope)
```

From what I understand, `evaluate` and other methods that need a `scope` aren't in control of constructing a `scope`, but access it through `setSafeProperty` and `getSafeProperty`.

### Proposal

I don't mind doing the work.

Since this is my first time contributing to mathjs, I'd rather file an issue, and agree a plan with you; rather than doing work which scratches my itch, but is unlikely to be merged.

1. `customs.js` defines some `getSafeProperty` and `setSafeProperty` which should be modified to recognized objects with `get` and `set` methods.
2. Add functions for `hasSafeProperty(object, prop)` and `getSafeProperties(object)`. Between these four functions we should be able to get and set, test membership, and enumerate entries.
3. Find and replace instances of `if (… in scope)` or `for (… in scope)` with the newly minted functions.
4. Add tests to `evaluate.test.js`, `simplify.test.js` and `security.test.js`.
5. Add an example of using a custom scope object.

Am I on the right track?

**Comment by josdejong** [comment]
Thanks for your clear proposal @jhugman . 

Just a first note: `getSafeProperty` and `setSafeProperty` are there for security reasons, they prevent for example accessing properties like `constructor` and `__proto__` on an object. I guess `Map` would be a secure alternative in that regard 😄 . At least, we need to be careful when making changes in those methods.

Before we decide whether we want to support `Map`: can you explain your use case? From you example, I understand you would like to have control over getting/setting variables in the scope?

**Comment by josdejong** [comment]
Thanks for making a clear case @jhugman. You fully conviced me that it is a great idea to implement support for the `Map` interface!

To me Approach 1 makes most sense (Approach 0 is something you can you yourself with a little wrapper function around `evaluate`). It will indeed open up a lot of flexibility, and I really like the security improvements that come with Map. I'm already thinking one step further: can we some day in the future completely drop support for plain objects in the evaluator in favor of Map? Plain objects are a security nightmare. But, one step at a time ;)

Some considerations:

1. We should check whether adding support for `Map` as an impact on the performance (ideally should not make evaluation noticeable slower)
2. We should implement support for Map using a ducktype interface checking. If we would check against `scope instanceof Map`, you can't provide a custom implementation of the `Map` interface like in your example above.

Help implementing this is very welcome!

**Comment by josdejong** [comment]
Published now in `v9.4.0`!
